File: frontend/src/api/snapshots.ts
/**
 * 议题显著性快照与跨国对比 API（详细设计 1.4 / 1.11）。
 *
 * 契约对齐后端 snapshots 路由：
 * - GET /api/v1/snapshots/topics/{id}?countries=CN,US&days=7
 *   响应 {topic_id, topic_name, timeline: {CC: [{window_start, article_count, salience_score, salience_rank}]}}
 *   本模块将其规整为 TopicSnapshot[]；路由不可用时回退 /topics/{id}/timeline 逐国拉取。
 * - GET /api/v1/snapshots/compare?countries=..&days=7[&topic_id=..]
 *   响应 {countries, days, per_country, disclaimer}（后端无 /topics/compare，不做错误回退）。
 */
import { ApiError, request } from "./client";

export interface SnapshotPoint {
  window_start: string;
  article_count: number;
  salience_score: number;
  salience_rank?: number | null;
  sentiment?: { pos: number; neu: number; neg: number };
}

export interface TopicSnapshot {
  topic_id: string;
  country_code: string;
  granularity: string;
  points: SnapshotPoint[];
}

/** 后端 /snapshots/compare 的 per_country 元素。 */
export interface CompareCountryPanel {
  country_code: string;
  salience_curve: { window_start: string; score: number; article_count: number }[];
  total_articles: number;
  top_topic_id: string | null;
  top_topic_name: string | null;
  coverage: "normal" | "low";
}

export interface TopicCompareResult {
  countries: string[];
  days: number;
  per_country: CompareCountryPanel[];
  disclaimer: string;
}

function isSnapshotUnavailable(err: unknown): boolean {
  // 路由不存在（404/HTTP 错误）或业务码 3001 视为“快照路由未上线”
  if (!(err instanceof ApiError)) return false;
  if (err.status === 404) return true;
  return err.code === 3001;
}

interface SnapshotsTimelineResponse {
  topic_id: string;
  topic_name: string;
  timeline: Record<
    string,
    { window_start: string; article_count: number; salience_score: number; salience_rank: number | null }[]
  >;
}

/** 获取议题在指定国家集合上的显著性快照。优先 snapshots 路由，回退 topics/timeline。 */
export async function fetchTopicSnapshots(
  topicId: string,
  countries: string[],
  days: number,
): Promise<TopicSnapshot[]> {
  const qs = new URLSearchParams({
    countries: countries.join(","),
    days: String(days),
  });
  try {
    const data = await request<SnapshotsTimelineResponse>(
      `/api/v1/snapshots/topics/${encodeURIComponent(topicId)}?${qs.toString()}`,
    );
    return Object.entries(data.timeline ?? {}).map(([cc, points]) => ({
      topic_id: data.topic_id,
      country_code: cc,
      granularity: "hour",
      points: points.map((p) => ({
        window_start: p.window_start,
        article_count: p.article_count,
        salience_score: p.salience_score,
        salience_rank: p.salience_rank,
      })),
    }));
  } catch (err) {
    if (!isSnapshotUnavailable(err)) throw err;
  }
  // 回退：逐国调 topics/{id}/timeline（后端该路由天然存在）
  const fallback = await Promise.all(
    countries.map(async (c) => {
      const q = new URLSearchParams({
        country_code: c,
        days: String(days),
        granularity: "day",
      });
      const data = await request<{
        topic_id: string;
        country_code: string | null;
        points: SnapshotPoint[];
      }>(`/api/v1/topics/${encodeURIComponent(topicId)}/timeline?${q.toString()}`);
      return {
        topic_id: data.topic_id,
        country_code: data.country_code || c,
        granularity: "day",
        points: data.points,
      };
    }),
  );
  return fallback;
}

/** 获取议题/全局跨国对比结果（含“统计关联≠因果”声明）。直接对接 /snapshots/compare。 */
export async function fetchTopicCompare(
  countries: string[],
  days: number,
  topicId?: string,
): Promise<TopicCompareResult> {
  const qs = new URLSearchParams({
    countries: countries.join(","),
    days: String(days),
  });
  if (topicId) qs.set("topic_id", topicId);
  return request<TopicCompareResult>(`/api/v1/snapshots/compare?${qs.toString()}`);
}

File: backend/app/api/routes/snapshots.py
"""agenda_snapshots 显著性时间线 + 跨国对比 API（T4.4）。"""
import uuid
from datetime import UTC, datetime, timedelta

from fastapi import APIRouter, Depends, Query
from sqlalchemy import select
from sqlalchemy.orm import Session

from app.api.deps import ROLE_AUTHORIZED, ROLE_REGISTERED, get_db, require_role
from app.core.errors import CODE_NOT_FOUND, BizError, ok
from app.models.snapshots import TopicSnapshot as AgendaSnapshot
from app.models.topic import Topic
from app.models.user import User

router = APIRouter()


@router.get("/topics/{topic_id}")
def topic_timeline(
    topic_id: uuid.UUID,
    countries: str | None = Query(None, description="逗号分隔，如 CN,US,JP"),
    days: int = Query(7, ge=1, le=90),
    db: Session = Depends(get_db),
    user: User = Depends(require_role(ROLE_REGISTERED)),
):
    topic = db.get(Topic, topic_id)
    if topic is None:
        raise BizError(CODE_NOT_FOUND, f"议题不存在: {topic_id}")

    window_start = datetime.now(UTC) - timedelta(days=days)
    stmt = select(AgendaSnapshot).where(
        AgendaSnapshot.topic_id == topic_id,
        AgendaSnapshot.window_start >= window_start,
    ).order_by(AgendaSnapshot.window_start.asc())
    if countries:
        cc_list = [c.strip().upper() for c in countries.split(",") if c.strip()]
        stmt = stmt.where(AgendaSnapshot.country_code.in_(cc_list))

    snaps = db.scalars(stmt).all()
    by_country: dict[str, list] = {}
    for snap in snaps:
        by_country.setdefault(snap.country_code, []).append({
            "window_start": snap.window_start.isoformat(),
            "article_count": snap.article_count,
            "salience_score": float(snap.salience_score or 0),
            "salience_rank": snap.salience_rank,
        })
    return ok({
        "topic_id": str(topic.id),
        "topic_name": topic.name,
        "timeline": by_country,
    })


@router.get("/compare")
def compare_countries(
    countries: str = Query(..., description="逗号分隔 ≤4 国"),
    days: int = Query(7, ge=1, le=90),
    topic_id: uuid.UUID | None = Query(None),
    db: Session = Depends(get_db),
    user: User = Depends(require_role(ROLE_AUTHORIZED)),
):
    cc_list = [c.strip().upper() for c in countries.split(",") if c.strip()]
    if len(cc_list) > 4:
        cc_list = cc_list[:4]
    window_start = datetime.now(UTC) - timedelta(days=days)

    per_country = []
    for cc in cc_list:
        stmt = select(AgendaSnapshot).where(
            AgendaSnapshot.window_start >= window_start,
            AgendaSnapshot.country_code == cc,
        ).order_by(AgendaSnapshot.window_start.asc())
        snaps = db.scalars(stmt).all()
        salience_curve = [
            {"window_start": s.window_start.isoformat(), "score": float(s.salience_score or 0), "article_count": s.article_count}
            for s in snaps
        ]
        total_articles = sum(s.article_count for s in snaps)
        top_topic_id = None
        top_topic_name = None
        if snaps:
            best = max(snaps, key=lambda s: float(s.salience_score or 0))
            top_topic_id = best.topic_id
            topic_obj = db.get(Topic, top_topic_id) if top_topic_id else None
            top_topic_name = topic_obj.name if topic_obj else None

        per_country.append({
            "country_code": cc,
            "salience_curve": salience_curve,
            "total_articles": total_articles,
            "top_topic_id": str(top_topic_id) if top_topic_id else None,
            "top_topic_name": top_topic_name,
            "coverage": "normal" if total_articles > 0 else "low",
        })

    return ok({
        "countries": cc_list,
        "days": days,
        "per_country": per_country,
        "disclaimer": "统计关联≠因果——显著性曲线仅反映不同国家对该议题的媒体报道相对强度，不能直接推断因果方向。",
    })


__all__ = ["router"]
